.github/workflows/erlang-library.yml

```diff
@@ -0,0 +1,164 @@
+
+env:
+  OTP_VERSION: 24.2
+  REBAR_VERSION: 3.18
+  THRIFT_VERSION: 0.14.2
+
+jobs:
+  deps:
+    name: Cache dependencies
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v2
+      
+      - name: Setup BEAM
+        uses: erlef/setup-beam@v1.10
+        with:
+          otp-version: ${{ env.OTP_VERSION }}
+          rebar3-version: ${{ env.REBAR_VERSION }}
+      
+      - name: Cache Hex packages
+        id: hex_cache
+        uses: actions/cache@v2
+        with:
+          path: ~/.cache/rebar3/hex/hexpm/packages
+          key: ${{ runner.os }}-otp-${{ env.OTP_VERSION }}-hex-${{ hashFiles('**/rebar.config') }}
+          restore-keys: |
+            ${{ runner.os }}-otp-${{ env.OTP_VERSION }}-hex-
+      
+      - name: Get dependencies
+        if: ${{ steps.hex_cache.outputs.cache-hit != 'true' }}
+        run: rebar3 deps
+
+  build:
+    name: Build / Check
+    runs-on: ubuntu-latest
+    needs: deps
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v2
+      
+      - name: Setup Thrift compiler
+        uses: valitydev/action-setup-thrift@v0.0.4
+        with:
+          thrift-version: ${{ env.THRIFT_VERSION }}
+      
+      - name: Setup BEAM
+        uses: erlef/setup-beam@v1.10
+        with:
+          otp-version: ${{ env.OTP_VERSION }}
+          rebar3-version: ${{ env.REBAR_VERSION }}
+      
+      - name: Cache Hex packages
+        uses: actions/cache@v2
+        with:
+          path: ~/.cache/rebar3/hex/hexpm/packages
+          key: ${{ runner.os }}-otp-${{ env.OTP_VERSION }}-hex-${{ hashFiles('**/rebar.config') }}
+          restore-keys: |
+            ${{ runner.os }}-otp-${{ env.OTP_VERSION }}-hex-
+      
+      - name: Compile
+        run: make compile
+
+      - name: Check formatting
+        run: make check_format
+      
+      - name: Run linting
+        run: make lint
+
+      - name: Run xref
+        run: make xref
+
+  analyze:
+    name: Dialyze
+    needs: deps
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v2
+      
+      - name: Setup Thrift compiler
+        uses: valitydev/action-setup-thrift@v0.0.4
+        with:
+          thrift-version: ${{ env.THRIFT_VERSION }}
+      
+      - name: Setup BEAM
+        uses: erlef/setup-beam@v1.10
+        with:
+          otp-version: ${{ env.OTP_VERSION }}
+          rebar3-version: ${{ env.REBAR_VERSION }}
+      
+      - name: Cache Hex packages
+        uses: actions/cache@v2
+        with:
+          path: ~/.cache/rebar3/hex/hexpm/packages
+          key: ${{ runner.os }}-otp-${{ env.OTP_VERSION }}-hex-${{ hashFiles('**/rebar.config') }}
+          restore-keys: |
+            ${{ runner.os }}-otp-${{ env.OTP_VERSION }}-hex-
+
+      - name: Cache PLTs
+        uses: actions/cache@v2
+        with:
+          path: |
+            ~/.cache/rebar3/rebar3_*_plt
+            _build/test/rebar3_*_plt
```

Comment:
Второй plt − это же расширение первого по сути? Есть ли смысл их оба кэшировать, если кэш всё равно сохраняется только (если я правильно понимаю) если все шаги успешны. Можно конечно job отдельный на сборку plt сделать как вариант, но выглядит накладно + непонятно сколько мы выиграем по времени в среднем, в живом репозитории с кэшами на мастере.

Reply:
Мне кажется есть, как минимум `~/.cache/rebar3/rebar3_*_plt` с мастер ветки в новом PR не надо будет собирать заново (а `_build/test/rebar3_*_plt` может реюзаться между билдами внутри одного PR).

Reply:
Ну просто даже с мастера всё равно либо оба этих файла приедут, либо ни одного. Я к тому, что не достаточно ли оставить только `_build/test/rebar3_*_plt`?

Reply:
> Ну просто даже с мастера всё равно либо оба этих файла приедут, либо ни одного.

Ну, да. Приедут два файла, базовый и расширенный. Базовый по дефолту содержит `[erts, crypto, kernel, stdlib]`, расширенный уже имеет зависимости приложения. Зачем каждый раз собирать базовый, когда надо только обновить расширенный?

Reply:
Мне казалось, что расширенный содержит `[erts, crypto, kernel, stdlib]` **и** все зависимости приложения. И соответственно, его одного хватит для всех последующих диализов в этом репозитории / PRе.
```
~/ws/vality/woody_erlang ❯ rebar3 dialyzer
===> Verifying dependencies...
===> Analyzing applications...
===> Compiling genlib
...
===> Analyzing applications...
===> Compiling woody
===> Dialyzer starting, this may take a while...
===> Updating plt...
===> Resolving files...
===> Updating base plt...
===> Resolving files...
===> Building with 205 files in /home/keynslug/.cache/rebar3/rebar3_24.2_plt...
===> Copying /home/keynslug/.cache/rebar3/rebar3_24.2_plt to /home/keynslug/ws/vality/woody_erlang/_build/default/rebar3_24.2_plt...
===> Checking 205 files in _build/default/rebar3_24.2_plt...
===> Adding 300 files to _build/default/rebar3_24.2_plt...
===> Doing success typing analysis...
===> Resolving files...
===> Analyzing 25 files with _build/default/rebar3_24.2_plt...

~/ws/vality/woody_erlang ❯ rm -v ~/.cache/rebar3/rebar3_24.2_plt
removed '/home/keynslug/.cache/rebar3/rebar3_24.2_plt'

~/ws/vality/woody_erlang ❯ rebar3 dialyzer
===> Verifying dependencies...
===> Analyzing applications...
===> Compiling woody
===> Dialyzer starting, this may take a while...
===> Updating plt...
===> Resolving files...
===> Checking 505 files in _build/default/rebar3_24.2_plt...
===> Doing success typing analysis...
===> Resolving files...
===> Analyzing 25 files with _build/default/rebar3_24.2_plt...

~/ws/vality/woody_erlang ❯ 
```

Reply:
Да, моей мыслью было скорее то, что будет, если между `rm -v ~/.cache/rebar3/rebar3_24.2_plt` и `rebar3 dialyzer` сделать, например, апгрейд зависимости -- я почему то думал, что расширенный plt инвалидируется целиком, но нет. 🍞🥐🥯